File: packages/core/src/grade.ts
/**
 * grade.ts — 答案判分逻辑
 *
 * 不同题型有不同的判分规则。容忍：空格、全/半角、中英文标点、数字格式。
 */

import type { Question } from "./types";

const TRUE_VALUES = new Set(["对", "正确", "true", "T", "✓", "√", "Y", "yes"]);
const FALSE_VALUES = new Set(["错", "错误", "false", "F", "✗", "×", "N", "no"]);

function normalize(s: string): string {
  return s
    .trim()
    .toLowerCase()
    .replace(/\s+/g, "")
    .replace(/，/g, ",")
    .replace(/。/g, ".")
    .replace(/（/g, "(")
    .replace(/）/g, ")")
    .replace(/：/g, ":");
}

/** 把分数 "3/4" 之类规范化 */
function normalizeNumeric(s: string): string {
  return normalize(s)
    .replace(/[^\d\-./%a-z]/g, "") // 去掉单位/汉字（保留数字、分数、百分号、字母用于"米"等）
    .replace(/^0+(\d)/, "$1");
}

export function gradeAnswer(question: Question, userAnswer: string): boolean {
  if (!userAnswer || !userAnswer.trim()) return false;
  const correct = question.answer;

  switch (question.type) {
    case "true_false": {
      const u = normalize(userAnswer);
      const c = normalize(correct);
      const userIsTrue = TRUE_VALUES.has(userAnswer.trim()) || TRUE_VALUES.has(u);
      const userIsFalse = FALSE_VALUES.has(userAnswer.trim()) || FALSE_VALUES.has(u);
      const correctIsTrue = TRUE_VALUES.has(correct.trim()) || TRUE_VALUES.has(c);
      if (userIsTrue || userIsFalse) return userIsTrue === correctIsTrue;
      return u === c;
    }

    case "choice": {
      // 答案可能是 "B"、"B. 60秒" 或直接存选项内容（如 "○○○○○○"）
      const u = userAnswer.trim().toUpperCase().charAt(0);
      let c = correct.trim().toUpperCase().charAt(0);
      // 若 answer 不是单个字母 A-D，则在 options 里反查下标
      if (!/^[A-D]$/.test(c) && question.options && question.options.length) {
        const cn = normalize(correct);
        const idx = question.options.findIndex(o => {
          const stripped = o.replace(/^[A-D][.、]\s*/, "");
          return normalize(o) === cn || normalize(stripped) === cn;
        });
        if (idx >= 0) c = String.fromCharCode(65 + idx);
      }
      return u === c;
    }

    case "fill_blank":
    case "calculation":
    case "word_problem": {
      // 多种容错：完全相等、normalize 相等、纯数字相等
      if (normalize(userAnswer) === normalize(correct)) return true;
      const un = normalizeNumeric(userAnswer);
      const cn = normalizeNumeric(correct);
      if (un && cn && un === cn) return true;
      // 数值容差（小数）
      const uf = parseFloat(un);
      const cf = parseFloat(cn);
      if (!isNaN(uf) && !isNaN(cf) && Math.abs(uf - cf) < 1e-6) return true;
      return false;
    }

    case "fill_blank_text": {
      // 文字填空（中文 / 英文单词）— 去空格、去大小写、去标点严格对比
      return normalizeText(userAnswer) === normalizeText(correct);
    }

    case "word_order": {
      // 用户答案应该是用逗号连接的词语序列
      const u = normalizeWordOrder(userAnswer);
      const c = normalizeWordOrder(correct);
      return u === c;
    }

    case "matching": {
      // 配对：A-1,B-2,C-3,D-4 顺序无关，集合相等即可
      const userPairs = parseMatchingAnswer(userAnswer);
      const correctPairs = parseMatchingAnswer(correct);
      if (userPairs.size !== correctPairs.size) return false;
      for (const [k, v] of userPairs) {
        if (correctPairs.get(k) !== v) return false;
      }
      return true;
    }

    default:
      return normalize(userAnswer) === normalize(correct);
  }
}

/** 文字填空规范化：trim、小写、去空格、去标点 */
function normalizeText(s: string): string {
  return s
    .trim()
    .toLowerCase()
    .replace(/\s+/g, "")
    .replace(/[，。！？：；,.!?:;()（）"'`]/g, "");
}

/** 排序题规范化：去掉空格，统一英文逗号 */
function normalizeWordOrder(s: string): string {
  return s.trim().replace(/，/g, ",").replace(/\s+/g, "");
}

/** 解析连线答案 "A-1,B-2,C-3,D-4" → Map { A: 1, B: 2, ... } */
function parseMatchingAnswer(s: string): Map<string, string> {
  const map = new Map<string, string>();
  const cleaned = s.trim().replace(/，/g, ",").replace(/\s+/g, "");
  if (!cleaned) return map;
  for (const pair of cleaned.split(",")) {
    const [k, v] = pair.split("-");
    if (k && v) map.set(k.toUpperCase(), v);
  }
  return map;
}

File: apps/mobile/ChinaTextbookStudy/Domain/Grade.swift
import Foundation

/// Answer grading — 1:1 port of packages/core/src/grade.ts.
///
/// Tolerances:
///   - whitespace, full-width vs half-width punctuation
///   - case-insensitive
///   - numeric formats (leading zeros, fractions, units)
///   - matching pairs are order-independent set equality
enum Grade {
    private static let trueValues: Set<String> = ["对", "正确", "true", "t", "✓", "√", "y", "yes"]
    private static let falseValues: Set<String> = ["错", "错误", "false", "f", "✗", "×", "n", "no"]

    static func gradeAnswer(question: Question, userAnswer: String) -> Bool {
        let trimmed = userAnswer.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return false }
        let correct = question.answer

        switch question.type {
        case .trueFalse:
            let u = normalize(userAnswer)
            let c = normalize(correct)
            let userIsTrue = trueValues.contains(trimmed) || trueValues.contains(u)
            let userIsFalse = falseValues.contains(trimmed) || falseValues.contains(u)
            let correctIsTrue = trueValues.contains(correct.trimmingCharacters(in: .whitespacesAndNewlines))
                || trueValues.contains(c)
            if userIsTrue || userIsFalse {
                return userIsTrue == correctIsTrue
            }
            return u == c

        case .choice:
            let u = String(trimmed.uppercased().first ?? " ")
            var c = String(correct.trimmingCharacters(in: .whitespacesAndNewlines).uppercased().first ?? " ")
            // If `answer` is not a single A-D letter, look it up in options.
            let isLetter = c.count == 1 && (c >= "A" && c <= "D")
            if !isLetter, !question.options.isEmpty {
                let cn = normalize(correct)
                let idx = question.options.firstIndex { opt in
                    let stripped = stripOptionPrefix(opt)
                    return normalize(opt) == cn || normalize(stripped) == cn
                }
                if let idx, idx < 4 {
                    c = String(UnicodeScalar(65 + idx)!)
                }
            }
            return u == c

        case .fillBlank, .calculation, .wordProblem:
            if normalize(userAnswer) == normalize(correct) { return true }
            let un = normalizeNumeric(userAnswer)
            let cn = normalizeNumeric(correct)
            if !un.isEmpty, !cn.isEmpty, un == cn { return true }
            if let uf = Double(un), let cf = Double(cn), abs(uf - cf) < 1e-6 { return true }
            return false

        case .fillBlankText:
            return normalizeText(userAnswer) == normalizeText(correct)

        case .wordOrder:
            return normalizeWordOrder(userAnswer) == normalizeWordOrder(correct)

        case .matching:
            let userPairs = parseMatchingAnswer(userAnswer)
            let correctPairs = parseMatchingAnswer(correct)
            return userPairs == correctPairs
        }
    }

    // MARK: - normalization helpers

    static func normalize(_ s: String) -> String {
        var out = s.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        out = out.components(separatedBy: .whitespacesAndNewlines).joined()
        let pairs: [(String, String)] = [
            ("，", ","), ("。", "."), ("（", "("), ("）", ")"), ("：", ":"),
        ]
        for (k, v) in pairs { out = out.replacingOccurrences(of: k, with: v) }
        return out
    }

    /// Strip leading "A. " / "B、" prefixes from option text.
    private static func stripOptionPrefix(_ s: String) -> String {
        guard let first = s.first, ("A"..."D").contains(first) else { return s }
        var idx = s.index(after: s.startIndex)
        if idx < s.endIndex {
            let c = s[idx]
            if c == "." || c == "、" { idx = s.index(after: idx) }
            while idx < s.endIndex, s[idx].isWhitespace { idx = s.index(after: idx) }
            return String(s[idx...])
        }
        return s
    }

    private static func normalizeNumeric(_ s: String) -> String {
        // keep digits, '-', '.', '/', '%' and ascii letters (e.g. "m" for meters)
        let allowed: Set<Character> = {
            var set = Set<Character>("0123456789-./%")
            for c in "abcdefghijklmnopqrstuvwxyz" { set.insert(c) }
            return set
        }()
        var out = normalize(s).filter { allowed.contains($0) }
        // strip leading zeros (but keep a single zero if everything is zeros)
        while out.count > 1, out.first == "0", let second = out.dropFirst().first, second.isNumber {
            out.removeFirst()
        }
        return out
    }

    private static func normalizeText(_ s: String) -> String {
        var out = s.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        out = out.components(separatedBy: .whitespacesAndNewlines).joined()
        let punct = Set<Character>("，。！？：；,.!?:;()（）\"'`")
        out.removeAll { punct.contains($0) }
        return out
    }

    private static func normalizeWordOrder(_ s: String) -> String {
        var out = s.trimmingCharacters(in: .whitespacesAndNewlines)
        out = out.replacingOccurrences(of: "，", with: ",")
        out = out.components(separatedBy: .whitespacesAndNewlines).joined()
        return out
    }

    private static func parseMatchingAnswer(_ s: String) -> [String: String] {
        var cleaned = s.trimmingCharacters(in: .whitespacesAndNewlines)
        cleaned = cleaned.replacingOccurrences(of: "，", with: ",")
        cleaned = cleaned.components(separatedBy: .whitespacesAndNewlines).joined()
        if cleaned.isEmpty { return [:] }
        var map: [String: String] = [:]
        for pair in cleaned.split(separator: ",") {
            let parts = pair.split(separator: "-", maxSplits: 1).map(String.init)
            if parts.count == 2, !parts[0].isEmpty, !parts[1].isEmpty {
                map[parts[0].uppercased()] = parts[1]
            }
        }
        return map
    }
}
